### PR TITLE
Add ONNX support for LayoutLMv3

### PR DESCRIPTION
**Pull request by regisss**
# What does this PR do?

This PR adds ONNX support for LayoutLMv3. Linked to #16308.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by regisss** [comment]
@lewtun All ONNX tests passed

File: src/transformers/models/layoutlmv3/configuration_layoutlmv3.py
Comment (NielsRogge) [inline src/transformers/models/layoutlmv3/configuration_layoutlmv3.py:215]: This is unfortunate, but we probably can't change it due to backwards compatibility.

cc @LysandreJik

Comment (lewtun) [reply src/transformers/models/layoutlmv3/configuration_layoutlmv3.py:215]: I wonder if we could add a deprecation warning to notify users that the order of arguments will change in v5? (Or is that type of breaking change too extreme?)

File: src/transformers/onnx/convert.py
Comment (lewtun) [inline src/transformers/onnx/convert.py:43]: Nice use of the type checks!

**Review by lewtun — APPROVED** [review]
Thanks a lot for adding another novel ONNX config @regisss - you're on 🔥 !

This PR LGTM, so gently pinging @LysandreJik or @sgugger for final approval 🙏 

Edit: I just saw that some of the CI tests failed. I've re-run them as they don't seem to be related to your changes. If they fail again, I suggest rebasing on `main` and pushing again to see if that resolves it. We should wait until the CI is green before merging ;-)

**Comment by regisss** [comment]
CI fails because of the following error:
```shell
Traceback (most recent call last):
  File "utils/check_repo.py", line 768, in <module>
    check_repo_quality()
  File "utils/check_repo.py", line 762, in check_repo_quality
    check_all_objects_are_documented()
  File "utils/check_repo.py", line 675, in check_all_objects_are_documented
    + "\n - ".join(undocumented_objs)
Exception: The following objects are in the public init so should be documented:
 - OptionalDependencyNotAvailable
 - dummy_scatter_objects
 - sys
```
But I do not see how my PR changes anything related to these objects.

File: src/transformers/models/layoutlmv3/configuration_layoutlmv3.py
Comment (sgugger) [inline src/transformers/models/layoutlmv3/configuration_layoutlmv3.py:22]: No absolute import here as this breaks everything. This should be imported from `...utils`. Also, it's only used for type checking, so use forward references in the type annotations and import this in an `if TYPE_CHECKING:` block.

File: src/transformers/models/layoutlmv3/configuration_layoutlmv3.py
Comment (sgugger) [inline src/transformers/models/layoutlmv3/configuration_layoutlmv3.py:27]: Same, should be imported for type checking only.

File: src/transformers/models/layoutlmv3/configuration_layoutlmv3.py
Comment (sgugger) [inline src/transformers/models/layoutlmv3/configuration_layoutlmv3.py:231]: Use forward references here.

**Review by sgugger — COMMENTED** [review]
That is what happens when you introduce new cyclical imports, it creates hard-to-debug errors everywhere ;-) 

So let's contain imports for type checking as forward references and not do top-level imports :-)

**Comment by regisss** [comment]
Wow thanks a lot @sgugger for the clear explanation, it makes complete sense!

**Comment by regisss** [comment]
CI and slow tests all passed. It should be ready now @sgugger @lewtun

**Comment by sgugger** [comment]
Thanks!

**Comment by githublsk** [comment]
@regisss Thank you for your great work, when convert layoutxlm  LayoutLMv2ForRelationExtraction to onnx, we are blocked by relation extraction layer for some reasons, can you try to export LayoutLMv2ForRelationExtraction model to onnx and give us for some help? gret thanks for you!

**Comment by githublsk** [comment]
@NielsRogge Thanks for your great work, when I convert LayoutLMv2ForRelationExtraction to onnx, I can not export relation extraction layer to onnx, can you help me to solve it? because the deadline is coming for the project, I hope you can help me, Thank you very much.

**Comment by regisss** [comment]
> @regisss Thank you for your great work, when convert layoutxlm LayoutLMv2ForRelationExtraction to onnx, we are blocked by relation extraction layer for some reasons, can you try to export LayoutLMv2ForRelationExtraction model to onnx and give us for some help? gret thanks for you!

@githublsk Where did you find `LayoutLMv2ForRelationExtraction`? I cannot find it in Transformers

**Comment by githublsk** [comment]
@regisss it is just in microsoft/unlim,please refer to the link:
https://github.com/microsoft/unilm/blob/master/layoutlmft/layoutlmft/models/layoutlmv2/modeling_layoutlmv2.py
image
it is useful for relation extraction, but when running onnx,some question occur as below:
image
the onnx graph is as below:
image
I can not find any reason, which confused me, I meet the deadline for my project, it is so urgent....

**Comment by githublsk** [comment]
@regisss if you have time, please help us, I am a newer to it, great thanks for you!

**Comment by regisss** [comment]
> @regisss if you have time, please help us, I am a newer to it, great thanks for you!

@githublsk Open an issue because it is not related to this PR. And provide the command/script you ran with the complete error message please, screenshots are not very helpful.

**Comment by githublsk** [comment]
@regisss Thank you for your great help, I open an issue in the link, can you help me? because the deadline is coming, it bothers me a lot, we hope you can help us to reslove it, great thanks.

https://github.com/huggingface/transformers/issues/17999

**Comment by gjj123** [comment]
> @regisss it is just in microsoft/unlim,please refer to the link: https://github.com/microsoft/unilm/blob/master/layoutlmft/layoutlmft/models/layoutlmv2/modeling_layoutlmv2.py image it is useful for relation extraction, but when running onnx,some question occur as below: image the onnx graph is as below: image I can not find any reason, which confused me, I meet the deadline for my project, it is so urgent....

how did you solve onnx convert error:
 **Exporting the operator bilinear to ONNX opset version 12 is not supported**

**Comment by pbcquoc** [comment]
@gjj123 replace torch.nn.Bilinear with this one
```
class Bilinear(nn.Module):
    def __init__(self, in1_features, in2_features, out_features):
        super(Bilinear, self).__init__()
        self.weight = torch.nn.Parameter(torch.zeros((in1_features, in2_features, out_features)))
        self.bias = torch.nn.Parameter(torch.zeros((out_features)))
        nn.init.xavier_uniform_(self.weight)

    def forward(self, x, y):
        t = x @ self.weight.permute(2,0,1)
        output = (t * y).sum(dim=2).t()
        return output
```

**Pull request merged**
